### PR TITLE
Rewrite OS X hardware events to use IOKit proper

### PR DESCRIPTION
**Pull request by theopolis**
This implements #1572. The IOKitHID publisher was rather limited in that it only reported a small subset of device attaches/detaches. This uses the basic IOKit attach/detach events but specifically handles PCI, Thunderbolt, and USD devices.

### CODE REVIEW

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

**Comment by theopolis** [comment]
Hey @sharvilshah, if you have a second and don't mind patching/picking to my branch, this could use another set of eyes. You have a bit of experience with the RunLoop-based APIs and can provide an honest review of the approach here.

**Comment by sharvilshah** [comment]
Nice, you already implemented this. I should be able to get to it sometime tomorrow.

File: osquery/events/darwin/iokit.cpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    return;
+  }
+
+  // Remove any existing stream.
+  stop();
+
+  port_ = IONotificationPortCreate(kIOMasterPortDefault);
+  // Get a run loop source from the created IOKit notification port.
+  auto run_loop_source = IONotificationPortGetRunLoopSource(port_);
+  CFRunLoopAddSource(run_loop_, run_loop_source, kCFRunLoopCommonModes);
```
Comment (sharvilshah) [inline osquery/events/darwin/iokit.cpp]: Just wondering what the difference between `kCFRunLoopDefaultMode` and `kCFRunLoopCommonModes` is. My google-fu seems weak currently, but it seems `kCFRunLoopDefaultMode` is a subset of `kCFRunLoopCommonModes`.

File: osquery/events/darwin/iokit.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+struct IOKitSubscriptionContext : public SubscriptionContext {
+  std::string model_id;
+  std::string vendor_id;
+
+  /// Bus type, e.g., USB.
+  std::string type;
+
+  /// Usage types.
+  std::string primary_usage;
+  std::string device_usage;
```
Comment (sharvilshah) [inline osquery/events/darwin/iokit.h]: I think these are not being used anymore, right?

**Comment by sharvilshah** [comment]
This looks great!

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

**Comment by theopolis** [comment]
Thanks for the help @sharvilshah!

**Pull request merged**